### DISCUSSION TITLE
Qual Framework JavaScript você usaria para iniciar um novo projeto?

### DISCUSSION THREAD

**Discussion by woliveiras** [frontendbr/forum#404] [Discussão]
Uma pergunta genérica para reunir o maior número de opiniões, possíveis!

Na issue #403 eu questionei sobre o uso de Angular 1.x em um projeto novo.

Gostaria de saber, agora, qual **Framework JavaScript vocês utilizariam para iniciar um projeto do zero hoje em dia? Por que?** - isso é uma pergunta seca mesmo para não enviesar a resposta, somente para levantar pontos de vista

**Comment by david27alves** [comment]
Dando uma pesquisada tenho visto a galera recomendando muito o Vue.js, pelo fato da semelhança que tem com AngularJS. Que mudou muito da versão 1.x para a 2.

**Comment by willycamargo** [comment]
IMHO, a melhor resposta seria: depende.

Nos últimos três projetos que trabalhei (sistemas) utilizamos:

1. **MithrilJS**
2. **ReactJS**
3. **ReactJS**

Mas eu confesso que não tenho uma opinião formada sobre qual framework utilizar sem conhecer as especificações do projeto e a equipe que irá dar manutenção nele.

Essa escolha esta muito ligada a esses aspectos e fazer uma escolha sem se basear no contexto certamente me faria escolher errado. Principalmente porque quero estar sempre aprendendo algo diferente e em cada projeto eu escolheria um framework diferente hahaha.

Apesar de trabalhar muito com **ReactJS** eu gostei muito do que vi no **Ember 2**, **Angular 2** e **VueJS**. E também fico de olho em algumas micro-libraries como **InfernoJS** e **Preact**

**Reforçando**: eu não escolheria o martelo antes de saber o tamanho do prego e onde ele será fixado 😅

**Comment by kakuzops** [comment]
Vou dar uma resposta diferente das outras aqui e será um pouco sem "expertise", eu atualmente sou estagiário e estudo muito sobre essas novas libs de javascript (react, vue, etc). Eu iniciaria um projeto com o  vue.js, por ele ser rápido por ele ter uma galera da comunidade muito ativa e sempre implementando coisa nova e tudo mais acho que ele futuramente ficaria muito fácil de dar manutenção. Mas não descartaria o React para começar um projeto do zero.

**Comment by luizbills** [comment]
vou responder como **qual Framework JavaScript vocês QUEREM utilizar para iniciar um projeto do zero hoje em dia? Por que?**

No front-end, eu estou querendo fazer algo com [choo](https://github.com/yoshuawuyts/choo) por causa da  API simples dele, por ser pequeno e acaba sendo fácil substituir ele futuramente por ter templates parecido com jsx.

Para back-end, eu estou querendo aplicar o [Serverless Framework](https://serverless.com/). Esse ano quero diminuir cada vez mais a necessidade de backend.

**Comment by expalmer** [comment]
[Inferno](https://github.com/infernojs/inferno). Motivo: React Like e muito rápido.

**Comment by roliveira-com** [comment]
Há muitos aspectos se considerar pra tomar essa decisão, mas hoje em dia eu iniciaria com o ReactJS pela estabilidade e escalabilidade.

**Comment by MaizerGomes** [comment]
> Bem, eu sou da área de back-end

Bem, eu sou da área de back-end mas ultimamente estou virado para aprender um pouco de front, especificamente para os frameworks js que ajudam muito na organização do projecto, performance da app e no UX, além de que gosto sempre de estar a par do que acontece nas linguagens que tenho conhecimento.

> Por um tempo tentei o **AngularJS 1**

Por um tempo tentei o **AngularJS 1** porque gostei muito da coisa de views, controllers e services/models, tudo separado, assemelhando-se muito à filosofia usada no back na maioria dos frameworks.

> Typescript

Tudo bem até que comecei a ler sobre diversos problemas que existiam no frame e lançaram a **v2** que era tudo diferente e com _Typescript_ por defeito. Como eu disse eu não sou front e aprender um novo sintaxe não era muito atraente naquela época. Eu queria algo fácil e rápido.

> **ReactJS** e **EmberJS**

Baixei muito material sobre **ReactJS** e **EmberJS** mas nunca cheguei a me sentir motivado a estudar ou mesmo a usar como fui com **Angular**. Para mim ele era perfeito mas muito mimi mi da internet me deixou intimidado e decidi aguardar por uma versão mais estável do **Angular2** com poucos bugs e backwards incompatibilities (_que nunca mais chega_).

> Um tal **VueJS**.

Desisti um pouco de aprender frames JS e me foquei no back-end de novo, porque as coisas andam muito renhidas ultimamente, até que comecei a ouvir um grande barulho por causa de um tal **VueJS**.
Ignorei durante um bom tempo até chegar ao ponto de não conseguir mais ignorar.

> `single-file-components`

Comecei a ver alguns tutoriais na net e percebi que era algo muito interessante apesar de algumas coisas eu não concordar como por exemplo o tão bem falado `single-file-components` (Pra quê ter HTML, JS e CSS em um único file???? Arghh..).

> talvez acho isso devido à forma diferente como as coisas são feitas no back-end

Apesar de algumas coisas que eu considero a forma como são feitas em **vueJS** de esquisitas (_talvez acho isso devido à forma diferente como as coisas são feitas no back-end_), este tem muita coisa divertida também: como o **two-way-data-binding**, os **components**(onde vc usa `<meu-componente></meu-componente>` e ele transforma para vc), a forma como os eventos são tratados(`@meu-evento="handler"`), a sintaxe(`@submit.prevent="formHandler"`) e tantas outras coisas que ainda estou a aprender.

> vue-bulma

Esta semana decidi por a mão na massa e usando um boilerplate admin([https://github.com/vue-bulma](https://github.com/vue-bulma)) que faz o uso de um framework css relativamente novo ([Bulma](http://bulma.io/)) comecei a redesenhar um monstro feito em 2012 que usa boostrap 2.3, js e PHP na mão (`cada linha de código que leio desse sistema dá vontade de pedir demissão e me atirar de uma ponte, mas hey, trabalho que é trabalho tem que sujar as mãos` 💯 ).

> viagem agradável

Até ao momento está a ser uma viagem agradável 🚆 (apesar de ainda não ter feito nem 2% do trabalho ainda) mas ainda estou a lutar para entender claramente aonde fica oquê, lidando com tutoriais desactualizados, webpack, babel, ES5/ES6, o próprio javascript e todos aqueles problemas de aprender algo novo.


Em suma, eu voto no **VueJS** porque é activamente desenvolvido, é fácil de pôr a correr, pode ser usado para criar um SPA ou não, possui uma documentação razoável, diversos plugins na comunidade,, vários tutoriais e demos na net (apesar de alguns estarem em mandariam, parece que é desenvolvido por um chinês).


> Incrível repositório com tudo que vc precisa para se tornar um Vuer:
[https://github.com/vuejs/awesome-vue](https://github.com/vuejs/awesome-vue)

![2017-01-10](https://cloud.githubusercontent.com/assets/10864249/21785472/57f6c2b2-d6c8-11e6-84e4-b24e43d9e567.png)

**Comment by alanhr** [comment]
Eu uso React + redux, porém já usei angular 2 até gostei mas me decepcionei com a forma de fazer os testes nele.

**Comment by deividmarques** [comment]
estou aprendendo vue.js e até o momento tenho gostado. Vale a pena uma leitura e o estudo sobre esse framework

**Comment by woliveiras** [comment]
Qual foi a parte que achou mais ruim nos testes com Angular 2, @alanhr?

**Comment by alanhr** [comment]
@woliveiras então achei a parte de teste com injeção de dependência  muito verboso.
Exemplo : quando você tem um componente que precisa usar uma classe de serviço, é necessário fazer uma config enorme  no beforeach do teste para aí sim você começar a testar.

**Comment by danvitoriano** [comment]
Bacana essa thread @woliveiras, mas apenas para fim de esclarecimento, vi muita gente sugerindo o uso de React, porém, React não é um framework, é apenas uma biblioteca! 

O Angular sim é um framework, completo, com tudo que alguém precisa pra criar um projeto front-end do zero.

Se você já está familiarizado com a sintaxe do Angular 2, esta seria uma boa opção para iniciar um projeto, usando principalmente o Angular CLI, com testes, tudo bonitinho.

Se você não tem a necessidade de um framework, o React é uma excelente escolha para lidar com interfaces componentizadas. Mas se você não vai criar componentes, nada melhor do que o velho HTML, CSS, JavaScript e um jQuery para fazer praticamente tudo que você precisa. :)

**Comment by caioincau** [comment]
Tenho usado fortemente o Vue.js no trabalho, para mim ele oferece um ótimo equilibrio entre o Angular e o React, além de ser mais performático.

https://vuejs.org/v2/guide/comparison.html

**Comment by matheusml** [comment]
@danvitoriano 

> React não é um framework, é apenas uma biblioteca!

Sim e não. Existe de fato uma _library_ chamada React, que é apenas isso, uma _library_ voltada  para a construção da sua UI.
Porém, geralmente quando a gente compara React com Angular, o que nos referimos é **ecossistema** React comparado com o **ecossistema** Angular.
E nesse caso, ambos são _frameworks_ completos para construir aplicações _web_.


@woliveiras 
Em relação a pergunta do tópico, eu acho que ela não faz sentido a menos que tenhamos um **contexto**. 
- Qual é a sua aplicação? 
- Quais são os problemas do cliente? 
- O que é aceitável e o que não é?
- Qual a performance esperada?

Somente depois de responder essas perguntas a gente vai ter o conhecimento necessário pra saber qual _framework_ escolher, e melhor: se de fato é necessário um _framework_.

**Comment by woliveiras** [comment]
Da hora @matheusml!

No caso eu quis ser bem genérico na pergunta pra gerar mais discussões quanto aos benefícios de cada ferramenta mesmo.

Meio que fazer um prós e contras, porém com cases da galera da comunidade.

Você tem alguns cases que fez com algumas dessas ferramentas que poderia compartilhar a experiência?

**Comment by ghost** [comment]
Meu Workflow básico para iniciar um projeto na parte da view é 

- pug.js
- sass
- gulp 
- browserSync

montei um gulp task com ele : https://github.com/rogeriorioli/gulp-pug-sass-starter

**o meu preferido é** 

- angular 1.x 
- pug.js
- sass
- gulp 
- browserSync

>  Claro que depende de cada projeto, quando faço uma landingPage esse workflow é perfeito , como aqui no trabalho usamos muito c# mvc 5 uso apenas o sass compilando com gulp

**Comment by hjdesigner** [comment]
Também sou do time "Depende do que seu projeto precisa".

Comecei um projeto esse mês com:
- gulp
- sass
- NodeJs

**Comment by laurindo** [comment]
Acho que independente de framework, você deve construir uma arquitetura
fácil de encaixar com qualquer outro framework. Eu estava vendo um conceito
muito legal chamado javascript isomorfico. Sugiro que procurem e tirem suas
conclusões.

Hoje em dia a briga entre angular 1, Angular 2, react, vue é grande. Qual
usar? Acredito que um código fácil de se adaptar ao framework é o melhor
caminho.

**Comment by matheusml** [comment]
@woliveiras 
> Você tem alguns cases que fez com algumas dessas ferramentas que poderia compartilhar a experiência?

Vamos lá.

De 2013 até 2015 trabalhei (quase que) exclusivamente com AngularJS.
Eu acreditava tanto no _framework_, e gostei tanto da revolução que vinha acontecendo no cenário JavaScript, que acabei migrando do _Back_ pro _Front_.
Trabalhei em 5 ou 6 projetos com AngularJS, de rede sociais até aplicações para pagamentos com alto nível de segurança requerido.
A minha stack padrão com Angular era:
- AngularJS 1.x 
- ES5
- Bower
- Grunt
- Jasmine
- JSHint

Até que, pela experiência adquirida, comecei a observar diversos problemas estruturais no _framework_, principalmente com o crescimento das aplicações.

A partir de 2016, comecei a trabalhar exclusivamente com React, também em projetos médios/grandes,
Diversos dos problemas que eu observava no ecossistema do Angular, foram resolvidos com maestria pelo React.
A stack padrão do React que uso:
- React / ReactDOM
- ES6
- React Router
- Webpack
- Eslint
- Mocha
- CSS Modules

Hoje estou mais do que satisfeito com o ecossistema do React, mas observo atentamente as alternativas, principalmente Elm e Angular 2.

**Comment by kakuzops** [comment]
Eu fico bem em duvida onde me enfiar pra aprender, já dei uma olhada  no Angular 1 com um tutorial sobre Angular para Zumbis, isso faz um bom tempo já, na época nem entendia ao certo como isso funcionava. Hoje que entendo mais o meio e tudo mais fico me perguntando se vale apena ainda aprender o Angular 1 ou aprender o react ou até mesmo aprender Angular 2. Fico receoso de me esforçar para aprender alguma tecnologia nova e no final não usa-la. Vejo muitas vagas de estagio( que é o meu perfil no momento) pedindo milhares de coisas, ai fico mais em duvida ainda. Espero que alguém me de uma luz e uma ajuda nessas coisas todas.

**Comment by laurindo** [comment]
Leonardo, pelo que vi você não está indo a fundo por conta do medo, hoje
angular e react são dois framework muito utilizados. Usa um e faz algo
grande  um projeto legal. Aprende tudo. Não fica com receio. Com certeza
você perde mais tempo, começando pedaço de cada um e não aprendendo nenhum.
E outra, independente de framework  seja ninja no javascript.

Espero ter ajudado.

**Comment by kakuzops** [comment]
@laurindo sim tenho medo de "aprender" e nunca usar para nada isso que complica mais, vou iniciar um estudo com o VUE porque achei ele sensacional e me senti motivado a aprende-lo, ser ninja no java script 
in progress.

**Comment by laurindo** [comment]
Leonardo, cara, eu indicaria você estudar react ao invés de angular 2. Mas
claro, javascript você deve ter um ótimo domínio.

No udemy tem um cursos muito bons de react + redux. Se mora aqui em
Fortaleza, o Milfont tem um curso de react.

Vue eu ficaria como terceira opção. Até agora não conheço empresas ou
amigos trabalhando com vue.js valendo mesmo não.

E pra complementar os estudos, estude inglês, porque lá fora o mercado tá
sedento por desenvolvedores, frontend, fullstack, devOps, web Developer,
etc.

**Comment by woliveiras** [comment]
@leonard9, nesse caso é melhor que veja se vai conseguir uma oportunidade de trampo em sua região com a ferramenta que você for estudar. Se for freela, sucesso!

Além do mais a dica do @laurindo é fato: aprenda JavaScript, Programação Orientada a Objetos, Programação Funcional, Padrões de Projetos, TDD, etc. Aprenda os conceitos de desenvolvimento de software e qualquer framework será pra você o que realmente deveria ser, somente uma ferramenta pra agilizar o que você já sabe fazer.

E o pior, mano, qualquer framework de hoje, pode ser o legado de amanhã. Você nunca vai aprender ou utilizar somente uma ferramenta.

Até mesmo se estiver trabalhando somente com coisa nova, mais cedo ou tarde você vai precisar dar manutenção em um legado. kkkkk

**Comment by wellingtonpgp** [comment]
Vue.js Vlws Flws 
Porque ele da conta do recado, simples e minimalista.  E galera do laravel utiliza muito ele, está crescendo bastante no meio dev.

**Comment by kakuzops** [comment]
@woliveiras @laurindo  Obrigado pelas dicas, vou fazer justamente isso, melhorar meu js ser ninja jedi do role kkk

**Comment by rmalveis** [comment]
Estou pra começar um projeto e optei pra aprender (pelo menos) 1 framework novo. 
Dois que me chamaram a atenção são o [Inferno Js](https://infernojs.org/) (ainda não mexi muito com React Like frameworks, então achei uma boa oportunidade e ele é rápido - pelos benchmarks!), que o @lnfnunes comentou e durante minhas pesquisas o [Cerebral JS](http://www.cerebraljs.com/) apareceu com um potencial muito interessante!

Farei algumas POCs para validação e, se agradarem mesmo, seguirei com elas!

**Comment by AbraaoAlves** [comment]
[AureliaJS](http://aurelia.io/) com certeza.

Se você precisa de um argumento forte para entender minha opinião, aqui vai:

A história da web nos ensina que só sobrevive quem está ligada aos padrões, ex: [EcmaScript4](http://stackoverflow.com/questions/2329602/why-was-ecmascript-4th-edition-completely-scrapped). Podem falar bem o quanto quiserem do JSX, CoffeeScript, Dart, ActionScript, ... Mas no final quem sobrevive é o padrão. Identificar tecnologia alinhadas nesse tipo de pensamento pode te dar uma pista de quem pode ter mais tempo de vida. Não se trata de popularidade e sim padronização. 

> Ex: Coffeescript já foi o preprocessador mais usado pela comunidade

#### _E qual a relação disso tudo com Aurelia?_

> [Learning Aurelia is just learning modern Javascript](https://ilikekillnerds.com/2016/12/give-aurelia-chance-2017/)

Isso não é apenas um ponto de convencimento, isso é um pilar do projeto: [Web Standards Focused](http://aurelia.io/hub.html#/doc/article/aurelia/framework/latest/what-is-aurelia/3)


E você: Qual a ideologia/princípio você mais se identifica/gosta no framework/library que você?
